python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -217,24 +107,51 @@ class IdentityPartitionMapping(PartitionMapping, NamedTuple("_IdentityPartitionM
     partitions in the downstream asset to the same partition in the upstream asset.
     """
 
-    def get_upstream_partitions_for_partition_range(
+    def get_upstream_mapped_partitions_result_for_partitions(
         self,
-        downstream_partition_key_range: Optional[PartitionKeyRange],
-        downstream_partitions_def: Optional[PartitionsDefinition],
+        downstream_partitions_subset: Optional[PartitionsSubset],
         upstream_partitions_def: PartitionsDefinition,
-    ) -> PartitionKeyRange:
-        if downstream_partitions_def is None or downstream_partition_key_range is None:
+        current_time: Optional[datetime] = None,
+        dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
+    ) -> UpstreamPartitionsResult:
+        if downstream_partitions_subset is None:
             check.failed("downstream asset is not partitioned")
 
-        return downstream_partition_key_range
+        if downstream_partitions_subset.partitions_def == upstream_partitions_def:
+            return UpstreamPartitionsResult(
+                downstream_partitions_subset,
+                [],
+            )
+
+        upstream_partition_keys = set(upstream_partitions_def.get_partition_keys())
+        downstream_partition_keys = set(downstream_partitions_subset.get_partition_keys())
+
+        return UpstreamPartitionsResult(
+            upstream_partitions_def.empty_subset().with_partition_keys(
+                list(upstream_partition_keys & downstream_partition_keys)
```

Comment:
Not critical right now, but at some point we might want to add an `__and__` method to `PartitionsSubset` so we can do these set operations without needing to convert back and forth between different collection types.

Reply:
Agree--we do lots of "and" operations in the asset backfill and partition status cache logic, so this would be helpful. Would prefer to add this in a separate PR though

Reply:
That's fine